File: src/main/java/sharev/domain/card/repository/CardRepository.java
package sharev.domain.card.repository;

import java.util.Optional;
import java.util.Set;
import java.util.UUID;
import org.springframework.data.jpa.repository.JpaRepository;
import sharev.domain.card.dto.response.ResponsePinNumberOnlyDto;
import sharev.domain.card.entity.Card;

public interface CardRepository extends JpaRepository<Card, Long>, CardRepositoryCustom {

    Optional<Card> findByGatheringIdAndAccountId(UUID gatheringId, Long accountId);

    Optional<Card> findByGatheringIdAndPinNumber(UUID gatheringId, Integer pinNumber);

    Set<ResponsePinNumberOnlyDto> findAllByGatheringId(UUID gatheringId);
}

File: src/main/kotlin/sharev/card/adapter/outbound/jpa/repository/CardRepository.kt
package sharev.card.adapter.outbound.jpa.repository

import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import org.springframework.data.jpa.repository.JpaRepository
import org.springframework.data.jpa.repository.Query
import sharev.card.adapter.outbound.jpa.entity.CardJpaEntity
import java.util.*

interface CardRepository : JpaRepository<CardJpaEntity, Long>,
    CardRepositoryCustom {
    fun findByGatheringIdAndAccountId(gatheringId: UUID, accountId: Long): CardJpaEntity?

    fun findByGatheringIdAndPinNumber(gatheringId: UUID, pinNumber: Int): CardJpaEntity?

    fun existsByGatheringIdAndAccountId(gatheringId: UUID, accountId: Long): Boolean

    fun findByAccountId(accountId: Long, pageable: Pageable): Page<CardJpaEntity>

    @Query("select c.pinNumber from CardJpaEntity c where c.gathering.id = :gatheringId and c.pinNumber is not null")
    fun findPinNumbersByGatheringId(gatheringId: UUID): Set<Int>
}
